.github/workflows/python.yml

```diff
@@ -11,18 +11,79 @@ jobs:
     runs-on: ${{ matrix.os }}
 
     strategy:
+      # We could consider using tox only
       matrix:
-        os: [ubuntu-20.04, macos-latest, windows-2019 ]
+        os: [windows-2019]
+        python-version: [3.6, 3.7, 3.8, 3.9]
+        python-env: [tox, global]
+        include:
+          # we install boost from apt/homebrew on linux/mac, and boost-python's
+          # sonames on those platforms look like "libboost_python36.so". so we
+          # can only build particular versions on particular platforms.
+        - os: ubuntu-20.04
+          python-version: 3.8
+          python-env: tox
+        - os: ubuntu-20.04
+          python-version: 3.8
+          python-env: global
+        - os: macos-latest
+          python-version: 3.9
+          python-env: tox
+        - os: macos-latest
+          python-version: 3.9
+          python-env: global
+        - os: macos-latest
+          python-version: 3.9
+          python-env: homebrew
```

Comment:
is it important to build for both python 3.9 and 3.8? it seems it bloats this workflow file quite a lot, and I imagine it'll make CI take longer too.

I wonder the same thing about `tox`. What's the benefit of using `tox` here?
I get the impression a lot of it boils down to making sure it works to install the python module in a virtual environment. That could be done by *just* testing `setup.py` in one, right? If it works in a venv, presumably it works without one as well.

Reply:
In practice, it is probably okay to just run tests against python 3.6. It's possible for language features to change between actively-supported python versions, and testing against all of them is the brute-force way to make sure we don't get stung. The other strategy is to test against the oldest version and maintain discipline to be aware of deprecations. We can tweak the tests and linters to help with this discipline, but it is less foolproof.

As much as I think reducing overall CI time is crucial, I think there are much bigger targets. Getting rid of appveyor is probably the most important. It might also be worth it to build a `setup-boost` github action analogous to `setup-python`, so we can run more piecemeal jobs at less cost, using consistent boost versions.

It's also interesting to note that the `cibuildwheel` process (as in #4218) does separate builds on all combinations of platform * python version, which would be even more combinations than I've listed here (we invoke `cibuildwheel` from one arbitrary python version, and it dispatches other jobs). We should probably dig into this process and make sure there's not too much waste (e.g. the artifacts of the main libtorrent library should probably be reused on a given platform), or else don't run `cibuildwheel` on every PR.

The benefits of `tox` are
* being a python-based test script, so test invocation is more portable
* isolation and reproducibility of the test environment
* running python packaging as part of the test script, to expose packaging issues
* a well-supported ecosystem to do basic stuff with minimal overhead
* a pluggable environment to do fancy stuff (tox testenvs can be run in docker, to make the build process even more isolated and reproducible)

I've never seen another python project with venv-specific issues. Even python extensions normally just use the `distutils` build system so they work fine in venvs. It was a real surprise to find this setup exposed so many issues with `b2`.

What do you think about: python 3.6 only + tox only? (No testing against global python installations)

Reply:
> I've never seen another python project with venv-specific issues. Even python extensions normally just use the distutils build system so they work fine in venvs. It was a real surprise to find this setup exposed so many issues with b2.

Yeah, `b2` is primarily a build system for C++, so I'm not too surprised its python support isn't particularly sophisticated (similar to how distutils C++ support isn't particularly sophisticated). I haven't taken a deep look at this yet, I will see if I can come up with any simplifications.

> What do you think about: python 3.6 only + tox only? (No testing against global python installations)

Sounds reasonable to me.